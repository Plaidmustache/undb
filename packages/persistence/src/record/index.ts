export * from './record.query-repository'
