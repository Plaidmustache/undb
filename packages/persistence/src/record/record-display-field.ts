import type { Field, IRecordDTO, IRecordDisplayValues, IRecordValues } from "@undb/table"

export type DisplayFieldName = `$${string}`

/**
 * Create a display field name for a field.
 * @param field The field to create a display field name for.
 * @returns The display field name.
 */
export function createDisplayFieldName(field: Field): DisplayFieldName {
  return `$${field.id.value}`
}

/**
 * Check if a field name is a display field name.
 * @param name The field name to check.
 * @returns `true` if the field name is a display field name, `false` otherwise.
 */
export function isDisplayerFieldName(name: string): name is DisplayFieldName {
  return name.startsWith("$")
}

/**
 * Get the raw field name from a display field name.
 * @param displayFieldName The display field name.
 * @returns The raw field name.
 */
export function getRawFieldName(displayFieldName: DisplayFieldName): string {
  return displayFieldName.slice(1)
}

export function getRecordDTOFromEntity(entity: any): IRecordDTO {
  const id = entity.id
  const values: IRecordValues = {}
  const displayValues: IRecordDisplayValues = {}
  for (const [key, value] of Object.entries(entity)) {
    if (key === "id") continue

    if (isDisplayerFieldName(key)) {
      displayValues[getRawFieldName(key)] = value
      continue
    }

    values[key] = value
  }

  return {
    id,
    values,
    displayValues,
  }
}
