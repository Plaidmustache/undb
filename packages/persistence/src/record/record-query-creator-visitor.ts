import {
  ID_TYPE,
  RatingField,
  SelectField,
  type AutoIncrementField,
  type CreatedAtField,
  type CreatedByField,
  type Field,
  type IFieldVisitor,
  type IdField,
  type NumberField,
  type ReferenceField,
  type RollupField,
  type StringField,
  type TableDo,
  type UpdatedAtField,
  type UpdatedByField,
} from "@undb/table"
import type { QueryCreator, SelectExpression } from "kysely"
import type { IQueryBuilder } from "../qb"
import { JoinTable } from "../underlying/reference/join-table"
import { UnderlyingTable } from "../underlying/underlying-table"
import { getRollupFn } from "../underlying/underlying-table.util"

export class RecordQueryCreatorVisitor implements IFieldVisitor {
  constructor(
    private readonly qb: IQueryBuilder,
    private readonly table: TableDo,
    private readonly foreignTables: Map<string, TableDo>,
    private readonly visibleFields: Field[],
  ) {}

  private creator: QueryCreator<any> | null = null

  create() {
    // handle select fields
    const referenceFields = this.table.schema.getReferenceFields(this.visibleFields)
    for (const referenceField of referenceFields) {
      referenceField.accept(this)
    }
    return this.creator ?? this.qb
  }

  id(field: IdField): void {}
  autoIncrement(field: AutoIncrementField): void {}
  createdAt(field: CreatedAtField): void {}
  createdBy(field: CreatedByField): void {}
  updatedAt(field: UpdatedAtField): void {}
  updatedBy(field: UpdatedByField): void {}
  string(field: StringField): void {}
  number(field: NumberField): void {}
  rating(field: RatingField): void {}
  select(field: SelectField): void {}
  reference(field: ReferenceField): void {
    const foreignTable = this.foreignTables.get(field.foreignTableId)
    if (!foreignTable) {
      return
    }

    const rollupFields = field.getRollupFields(this.visibleFields)

    const underlyingForiegnTable = new UnderlyingTable(foreignTable)

    const joinTable = new JoinTable(this.table, field)
    const name = joinTable.getTableName()
    const valueField = joinTable.getValueFieldId()
    const symmetricField = joinTable.getSymmetricValueFieldId()

    const visible = this.visibleFields.some((f) => f.id.equals(field.id))
    const displayFields = visible ? foreignTable.schema.getDisplayFields() : []

    this.creator = (this.creator || this.qb).with(field.id.value, (db) =>
      db
        .selectFrom(name)
        .innerJoin(
          underlyingForiegnTable.name,
          `${name}.${symmetricField}`,
          `${underlyingForiegnTable.name}.${ID_TYPE}`,
        )
        .select(
          (sb) =>
            [
              `${name}.${valueField} as ${ID_TYPE}`,
              visible ? sb.fn("json_group_array", [sb.ref(`${name}.${symmetricField}`)]).as(field.id.value) : undefined,
              // select display fields for reference
              ...displayFields.map((f) =>
                sb.fn("json_group_array", [sb.ref(`${underlyingForiegnTable.name}.${f.id.value}`)]).as(f.id.value),
              ),
              ...rollupFields.map((rollupField) =>
                sb
                  .fn(getRollupFn(rollupField.fn), [
                    sb.ref(`${underlyingForiegnTable.name}.${rollupField.rollupFieldId}`),
                  ])
                  .as(rollupField.id.value),
              ),
            ].filter(Boolean) as SelectExpression<any, any>[],
        )
        .groupBy(`${name}.${valueField}`),
    )
  }
  rollup(field: RollupField): void {}
}
