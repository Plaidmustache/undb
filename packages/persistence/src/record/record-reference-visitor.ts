import {
  ID_TYPE,
  RatingField,
  SelectField,
  type AutoIncrementField,
  type CreatedAtField,
  type CreatedByField,
  type Field,
  type IFieldVisitor,
  type IdField,
  type NumberField,
  type ReferenceField,
  type RollupField,
  type StringField,
  type TableDo,
  type UpdatedAtField,
  type UpdatedByField,
} from "@undb/table"
import type { SelectQueryBuilder } from "kysely"

export class RecordReferenceVisitor implements IFieldVisitor {
  constructor(
    private qb: SelectQueryBuilder<any, any, any>,
    private readonly table: TableDo,
  ) {}

  join(fields: Field[]) {
    for (const field of fields) {
      field.accept(this)
    }
    return this.qb
  }

  id(field: IdField): void {
    throw new Error("Method not implemented.")
  }
  autoIncrement(field: AutoIncrementField): void {
    throw new Error("Method not implemented.")
  }
  createdAt(field: CreatedAtField): void {
    throw new Error("Method not implemented.")
  }
  createdBy(field: CreatedByField): void {
    throw new Error("Method not implemented.")
  }
  updatedAt(field: UpdatedAtField): void {
    throw new Error("Method not implemented.")
  }
  updatedBy(field: UpdatedByField): void {
    throw new Error("Method not implemented.")
  }
  string(field: StringField): void {
    throw new Error("Method not implemented.")
  }
  number(field: NumberField): void {
    throw new Error("Method not implemented.")
  }
  rating(field: RatingField): void {
    throw new Error("Method not implemented.")
  }
  select(field: SelectField): void {
    throw new Error("Method not implemented.")
  }
  reference(field: ReferenceField): void {
    this.qb = this.qb.leftJoin(field.id.value, `${this.table.id.value}.${ID_TYPE}`, `${field.id.value}.${ID_TYPE}`)
  }
  rollup(field: RollupField): void {
    throw new Error("Method not implemented.")
  }
}
