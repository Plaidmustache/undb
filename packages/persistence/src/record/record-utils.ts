import { createLogger } from "@undb/logger"
import {
  FieldIdVo,
  ID_TYPE,
  type Field,
  type IRecordDTO,
  type IRecordDisplayValues,
  type IRecordValues,
  type TableDo,
} from "@undb/table"
import { format } from "date-fns/fp"
import { isNumber, isString } from "radash"

const formatter = format("yyyy-MM-dd")

export type DisplayFieldName = `$${string}`

/**
 * Create a display field name for a field.
 * @param field The field to create a display field name for.
 * @returns The display field name.
 */
export function createDisplayFieldName(field: Field): DisplayFieldName {
  return `$${field.id.value}`
}

/**
 * Check if a field name is a display field name.
 * @param name The field name to check.
 * @returns `true` if the field name is a display field name, `false` otherwise.
 */
export function isDisplayerFieldName(name: string): name is DisplayFieldName {
  return name.startsWith("$")
}

/**
 * Get the raw field name from a display field name.
 * @param displayFieldName The display field name.
 * @returns The raw field name.
 */
export function getRawFieldName(displayFieldName: DisplayFieldName): string {
  return displayFieldName.slice(1)
}

/**
 * Returns the expanded field name for a given field.
 *
 * @param field - The field object.
 * @returns The expanded field name.
 */
export function getJsonExpandedFieldName(field: Field): string {
  return `${field.id.value}_expanded`
}

const logger = createLogger("record-utils")

/**
 * Get the record DTO from an entity. The record DTO is the DTO that is returned to the client. It contains the values and display values of the record.
 * @param entity The entity to get the record DTO from.
 * @returns The record DTO.
 */
export function getRecordDTOFromEntity(table: TableDo, entity: any, foreignTables: Map<string, TableDo>): IRecordDTO {
  const id = entity.id
  const values: IRecordValues = {}
  const displayValues: IRecordDisplayValues = {}
  const schema = table.schema
  for (let [key, value] of Object.entries(entity)) {
    if (key === ID_TYPE) continue
    const isDisplay = isDisplayerFieldName(key)
    if (isDisplay) {
      key = getRawFieldName(key as DisplayFieldName)
    }

    const field = schema.getFieldById(new FieldIdVo(key)).into(undefined)
    if (!field) continue

    if (isDisplay) {
      if (field.type === "reference" && isString(value)) {
        const parsed = JSON.parse(value)
        for (let [key, value] of Object.entries(parsed)) {
          parsed[key] = JSON.parse(value as string)
        }
        displayValues[key] = parsed
      } else if (
        (field.type === "user" || field.type === "createdBy" || field.type === "updatedBy") &&
        isString(value)
      ) {
        displayValues[key] = JSON.parse(value)
      } else if (value) {
        displayValues[key] = value
      }
      continue
    } else {
      if (field.type === "select") {
        if (field.isSingle) {
          const optionId = value
          const option = field.options.find((o) => o.id === optionId)
          displayValues[key] = option?.name
        } else if (Array.isArray(value)) {
          const optionIds = value
          displayValues[key] = optionIds.map((optionId) => {
            const option = field.options.find((o) => o.id === optionId)
            return option?.name
          })
        }
      }

      if (
        (field.type === "reference" ||
          field.type === "attachment" ||
          field.type === "json" ||
          ((field.type === "select" || field.type === "user") && field.isMultiple)) &&
        isString(value)
      ) {
        try {
          values[key] = JSON.parse(value)
        } catch (error) {
          logger.warn({ error, value }, "Error parsing JSON")
          values[key] = [value]
        }
        continue
      }

      if (field.type === "rollup" && field.fn === "lookup") {
        const foreignTable = field.getForeignTable(table, foreignTables)
        let lookupField: Field | undefined = undefined
        if (foreignTable) {
          const rollupField = field.getRollupField(foreignTable)
          if (rollupField.isSome()) {
            lookupField = rollupField.unwrap()
          }
        }

        function handleValue(value: (string | number)[]) {
          if (lookupField?.type === "date") {
            return value
              .map((v: string | number) => (v ? new Date(v).toISOString() : null))
              .map((v) => (v ? formatter(v) : null))
          } else if (lookupField?.type === "select") {
            return value.map((v) => lookupField.getOptionById(v)?.name ?? null)
          } else {
            return value
          }
        }

        if (Array.isArray(value)) {
          values[key] = handleValue(value)
          continue
        } else if (isString(value)) {
          try {
            values[key] = JSON.parse(value)
            values[key] = handleValue(values[key])
          } catch (error) {
            logger.warn({ error, value }, "Error parsing JSON")
            values[key] = [value]
          }
          continue
        }
      }

      if (field.type === "checkbox") {
        values[key] = Boolean(value)
        continue
      }
      if (field.type === "date" && (isString(value) || isNumber(value))) {
        values[key] = new Date(value).toISOString()
        continue
      }
      values[key] = value
    }
  }

  return {
    id,
    values,
    displayValues,
  }
}
