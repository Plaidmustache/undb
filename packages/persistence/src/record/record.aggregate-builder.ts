import type { Field, IFieldAggregate } from "@undb/table"
import { sql, type AliasedExpression, type ExpressionBuilder } from "kysely"
import { match } from "ts-pattern"
import type { UnderlyingTable } from "../underlying/underlying-table"

export class AggregateFnBuiler {
  constructor(
    private readonly table: UnderlyingTable,
    private readonly eb: ExpressionBuilder<any, any>,
    private readonly field: Field,
    private readonly aggregate: IFieldAggregate,
  ) {}

  public build(): AliasedExpression<any, any> {
    const alias = this.field.id.value
    const fieldId = this.table.getFieldName(alias)

    return match(this.aggregate)
      .returnType<AliasedExpression<any, any>>()
      .with("sum", () => this.eb.fn.sum(fieldId).as(alias))
      .with("avg", () => this.eb.fn.avg(fieldId).as(alias))
      .with("min", () => this.eb.fn.min(fieldId).as(alias))
      .with("max", () => this.eb.fn.max(fieldId).as(alias))
      .with("count", () => this.eb.fn.countAll().as(alias))
      .with("count_uniq", () => this.eb.fn.count(fieldId).distinct().as(alias))
      .with("count_empty", () => sql`COUNT(*) - COUNT(NULLIF(${sql.ref(fieldId)}, ''))`.as(alias))
      .with("count_not_empty", () =>
        sql`COUNT(CASE WHEN ${sql.ref(fieldId)} IS NOT NULL AND ${sql.ref(fieldId)} != 0 THEN 1 END)`.as(alias),
      )
      .with("percent_empty", () => sql`(COUNT(*) - COUNT(NULLIF(${sql.ref(fieldId)}, ''))) * 1.0 / COUNT(*)`.as(alias))
      .with("percent_not_empty", () => sql`COUNT(NULLIF(${sql.ref(fieldId)}, '')) * 1.0 / COUNT(*)`.as(alias))
      .with("percent_uniq", () => sql`COUNT(DISTINCT ${sql.ref(fieldId)}) * 1.0 / COUNT(*)`.as(alias))
      .with("count_true", () => sql`COUNT(CASE WHEN ${sql.ref(fieldId)} THEN 1 END)`.as(alias))
      .with("count_false", () => sql`COUNT(CASE WHEN NOT ${sql.ref(fieldId)} THEN 1 END)`.as(alias))
      .with("percent_true", () => sql`COUNT(CASE WHEN ${sql.ref(fieldId)} THEN 1 END) * 1.0 / COUNT(*)`.as(alias))
      .with("percent_false", () => sql`COUNT(CASE WHEN NOT ${sql.ref(fieldId)} THEN 1 END) * 1.0 / COUNT(*)`.as(alias))
      .exhaustive()
  }
}
