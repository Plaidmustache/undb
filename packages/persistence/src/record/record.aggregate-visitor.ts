import type {
  AutoIncrementField,
  CreatedAtField,
  IFieldAggregate,
  IFieldVisitor,
  IdField,
  NumberField,
  StringField,
  UpdatedAtField,
} from "@undb/table"
import { sql, type AliasedExpression, type ExpressionBuilder } from "kysely"
import { match } from "ts-pattern"

export class RecordAggregateVisitor implements IFieldVisitor {
  constructor(
    protected readonly eb: ExpressionBuilder<any, any>,
    private readonly aggregate: IFieldAggregate,
  ) {}

  #ebs: AliasedExpression<any, any>[] = []
  public get ebs() {
    return this.#ebs
  }

  id(field: IdField): void {
    // throw new Error("Method not implemented.")
  }
  autoIncrement(field: AutoIncrementField): void {
    // throw new Error("Method not implemented.")
  }
  createdAt(field: CreatedAtField): void {
    // throw new Error("Method not implemented.")
  }
  updatedAt(field: UpdatedAtField): void {
    // throw new Error("Method not implemented.")
  }
  string(field: StringField): void {
    const aggergate = field.aggregate.parse(this.aggregate)
    const eb = match(aggergate)
      .returnType<AliasedExpression<any, any>>()
      .with("count_uniq", () => this.eb.fn.count(field.id.value).distinct().as(field.id.value))
      .with("count_empty", () => sql`COUNT(*) - COUNT(NULLIF(${sql.ref(field.id.value)}, ''))`.as(field.id.value))
      .with("count_not_empty", () =>
        sql`COUNT(CASE WHEN ${sql.ref(field.id.value)} IS NOT NULL AND ${sql.ref(field.id.value)} != '' THEN 1 END)`.as(
          field.id.value,
        ),
      )
      .with("percent_empty", () =>
        sql`(COUNT(*) - COUNT(NULLIF(${sql.ref(field.id.value)}, ''))) * 100.0 / COUNT(*)`.as(field.id.value),
      )
      .exhaustive()

    this.#ebs.push(eb)
  }
  number(field: NumberField): void {
    // throw new Error("Method not implemented.")
  }
}
