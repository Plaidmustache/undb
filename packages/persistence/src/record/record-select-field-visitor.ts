import {
  AttachmentField,
  CheckboxField,
  DateField,
  ID_TYPE,
  JsonField,
  RatingField,
  SelectField,
  type AutoIncrementField,
  type CreatedAtField,
  type CreatedByField,
  type Field,
  type IFieldVisitor,
  type IdField,
  type NumberField,
  type ReferenceField,
  type RollupField,
  type StringField,
  type TableDo,
  type UpdatedAtField,
  type UpdatedByField,
} from "@undb/table"
import type { EmailField } from "@undb/table/src/modules/schema/fields/variants/email-field"
import { getTableName } from "drizzle-orm"
import { sql, type ExpressionBuilder, type SelectExpression } from "kysely"
import { users } from "../tables"
import type { UnderlyingTable } from "../underlying/underlying-table"
import { createDisplayFieldName } from "./record-utils"

export class RecordSelectFieldVisitor implements IFieldVisitor {
  #select: SelectExpression<any, any>[] = []

  addSelect(select: SelectExpression<any, any>): void {
    this.#select.push(select)
  }

  public $select(fields: Field[]): SelectExpression<any, any>[] {
    for (const field of fields) {
      field.accept(this)
    }
    return this.#select
  }

  private getField(field: string) {
    return `${this.table.name}.${field} as ${field}`
  }

  constructor(
    private readonly table: UnderlyingTable,
    private readonly foreignTables: Map<string, TableDo>,
    private readonly eb: ExpressionBuilder<any, string>,
  ) {
    this.addSelect(this.getField(ID_TYPE))
  }
  select(field: SelectField): void {
    this.addSelect(this.getField(field.id.value))
  }

  id(field: IdField): void {
    // this.addSelect(this.getField(field.id.value))
  }
  autoIncrement(field: AutoIncrementField): void {
    this.addSelect(this.getField(field.id.value))
  }
  createdAt(field: CreatedAtField): void {
    this.addSelect(this.getField(field.id.value))
  }
  createdBy(field: CreatedByField): void {
    this.addSelect(this.getField(field.id.value))
    const user = getTableName(users)
    const as = createDisplayFieldName(field)

    const name = this.eb
      .selectFrom(user)
      .select(`${user}.${users.username.name}`)
      .whereRef(field.id.value, "=", `${user}.${users.id.name}`)
      .limit(1)
      .as(as)

    this.addSelect(name)
  }

  updatedBy(field: UpdatedByField): void {
    this.addSelect(this.getField(field.id.value))
    const user = getTableName(users)
    const as = createDisplayFieldName(field)

    const name = this.eb
      .selectFrom(user)
      .select(`${user}.${users.username.name}`)
      .whereRef(field.id.value, "=", `${user}.${users.id.name}`)
      .limit(1)
      .as(as)

    this.addSelect(name)
  }
  updatedAt(field: UpdatedAtField): void {
    this.addSelect(this.getField(field.id.value))
  }
  string(field: StringField): void {
    this.addSelect(this.getField(field.id.value))
  }
  number(field: NumberField): void {
    this.addSelect(this.getField(field.id.value))
  }
  rating(field: RatingField): void {
    this.addSelect(this.getField(field.id.value))
  }
  email(field: EmailField): void {
    this.addSelect(this.getField(field.id.value))
  }
  json(field: JsonField): void {
    this.addSelect(this.getField(field.id.value))
  }
  reference(field: ReferenceField): void {
    const select = `${field.id.value}.${field.id.value} as ${field.id.value}`
    this.addSelect(select)

    const name = createDisplayFieldName(field)

    const foreignTable = this.foreignTables.get(field.foreignTableId)
    if (foreignTable) {
      const displayFields = foreignTable.schema.getDisplayFields()
      const select = this.eb
        .fn(
          "json_object",
          displayFields.flatMap((displayField) => [
            sql.raw(`'${displayField.id.value}'`),
            `${field.id.value}.${displayField.id.value}`,
          ]),
        )
        .as(name)

      this.addSelect(select)
    }
  }
  rollup(field: RollupField): void {
    const select = `${field.referenceFieldId}.${field.id.value} as ${field.id.value}`
    this.addSelect(select)
  }
  attachment(field: AttachmentField): void {
    this.addSelect(this.getField(field.id.value))
  }
  date(field: DateField): void {
    this.addSelect(this.getField(field.id.value))
  }
  checkbox(field: CheckboxField): void {
    this.addSelect(this.getField(field.id.value))
  }
}
