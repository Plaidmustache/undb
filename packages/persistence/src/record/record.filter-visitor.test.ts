import { Filter, Schema, type IFilterGroup } from '@undb/table'
import Database from 'bun:sqlite'
import { describe, expect, test } from 'bun:test'
import { Kysely } from 'kysely'
import { BunSqliteDialect } from 'kysely-bun-sqlite'
import type { IQueryBuilder } from '../qb'
import { RecordFilterVisitor } from './record.filter-visitor'

const schema = Schema.fromJSON([
  { id: 'field1', type: 'string', name: 'field1' },
  { id: 'field2', type: 'number', name: 'field2' },
])

const sqlite = new Database()
const qb = new Kysely({
  dialect: new BunSqliteDialect({
    database: sqlite,
  }),
}) satisfies IQueryBuilder

describe('record.filter-visitor', () => {
  test.each<IFilterGroup>([
    {
      conjunction: 'and',
      children: [
        { fieldId: 'field1', op: 'eq', value: 'value1' },
        { fieldId: 'field2', op: 'gt', value: 1 },
      ],
    },
    {
      conjunction: 'or',
      children: [
        { fieldId: 'field1', op: 'eq', value: 'value1' },
        {
          conjunction: 'and',
          children: [
            { fieldId: 'field1', op: 'eq', value: 'value1' },
            { fieldId: 'field2', op: 'gt', value: 1 },
          ],
        },
        {
          conjunction: 'or',
          children: [
            { fieldId: 'field1', op: 'eq', value: 'value2' },
            { fieldId: 'field2', op: 'lt', value: 2 },
          ],
        },
      ],
    },
  ])('should get query', (filter) => {
    const f = new Filter(filter)
    const spec = f.getSpec(schema)

    const query = qb
      .selectFrom('table')
      .selectAll()
      .where((eb) => {
        const visitor = new RecordFilterVisitor(eb)
        if (spec.isSome()) {
          spec.unwrap().accept(visitor)
        }

        return visitor.cond
      })
      .compile()

    expect(query.sql).toMatchSnapshot()
    expect(query.parameters).toMatchSnapshot()
  })
})
