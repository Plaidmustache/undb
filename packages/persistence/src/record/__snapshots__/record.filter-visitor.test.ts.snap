// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`record.filter-visitor should get query 1`] = `"select * from "table" where ("field1" = ? and "field2" > ?)"`;

exports[`record.filter-visitor should get query 2`] = `
[
  "value1",
  1,
]
`;

exports[`record.filter-visitor should get query 3`] = `"select * from "table" where (("field1" = ? or ("field1" = ? and "field2" > ?)) or ("field1" = ? or "field2" < ?))"`;

exports[`record.filter-visitor should get query 4`] = `
[
  "value1",
  "value1",
  1,
  "value2",
  2,
]
`;
